### PR TITLE
Template Instantiator Test Suite

### LINKED ISSUE #124: Test suite for template_instantiator
We should have a dedicated suite of tests for `template_instantiator` to make testing and debugging easier.

### PR DESCRIPTION
**Pull request by varunagrawal**
Fixes #124

### CODE REVIEW

File: tests/test_template_instantiator.py [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+            template<U={double}>
+            double method(const T x, const U& param);
+            """)[0]
+        cls = Class.rule.parseString("""
+            template<T={string}>
+            class Foo {};
+        """)[0]
+        typenames = ['T', 'U']
+        class_instantiations = [Typename.rule.parseString("string")[0]]
+        method_instantiations = [Typename.rule.parseString("double")[0]]
```
Comment (gchenfc) [inline tests/test_template_instantiator.py:63]: I'm kind of confused why these need to be explicitly defined; don't they get parsed by `original.template.instantiations` ?  Or rather, is it true that `class_instantiations` should be the same as `cls.template.instantiations` ?

File: gtwrap/template_instantiator.py [hunk 2]
```diff
@@ -201,11 +223,14 @@ class InstantiationHelper:
...
+    def __init__(self, instantiation_type: InstantiatedMembers):
         self.instantiation_type = instantiation_type
 
-    def instantiate(self, instantiated_methods, method, typenames,
-                    class_instantiations, method_instantiations, parent):
+    def instantiate(self, instantiated_methods: List[InstantiatedMembers],
+                    method: ClassMembers, typenames: Sequence[str],
+                    class_instantiations: Sequence[parser.Typename],
+                    method_instantiations: Sequence[parser.Typename],
+                    parent: 'InstantiatedClass'):
```
Comment (gchenfc) [inline gtwrap/template_instantiator.py:233]: Re: previous comment in unit test

In this function, why isn't `method_instantiations` the same as `method.template.instantiations` and `class_instantiations` the same as `parent.template.instantiations` ?

File: tests/test_template_instantiator.py [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+        self.assertIsInstance(self.func.original, GlobalFunction)
+        self.assertEqual(self.func.name, "functionIntDouble")
+        self.assertEqual(len(self.func.args.list()), 1)
+        self.assertEqual(self.func.args.list()[0].ctype.get_typename(), "int")
+        self.assertEqual(self.func.return_type.type1.get_typename(), "double")
+
+    def test_to_cpp(self):
+        """Test to_cpp method."""
+        actual = self.func.to_cpp()
+        self.assertEqual(actual, "function<int,double>")
```
Comment (gchenfc) [inline tests/test_template_instantiator.py:141]: I like these `to_cpp` functions/unit tests - don't know why I never noticed them before

File: tests/test_template_instantiator.py [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+
+        self.assertEqual(instantiated_return_type.type1.get_typename(),
+                         "double")
+        self.assertEqual(instantiated_return_type.type2.get_typename(), "char")
+
+    def test_instantiate_name(self):
+        """Test for instantiate_name."""
+        instantiations = [Typename.rule.parseString("Man")[0]]
+        instantiated_name = instantiate_name("Iron", instantiations)
+        self.assertEqual(instantiated_name, "IronMan")
```
Comment (gchenfc) [inline tests/test_template_instantiator.py:553]: lol

File: gtwrap/template_instantiator.py (hunk 2 above)
Comment (gchenfc) [reply gtwrap/template_instantiator.py:233]: *Edit: Ok I see what's going on now

Comment (varunagrawal) [reply gtwrap/template_instantiator.py:233]: Yeah I realized I was copy pasting the same pattern again and again so I just abstracted it out and tested it to make it iron clad 😁

**Pull request merged**